bson/src/main/org/bson/json/JsonStreamBuffer.java

```diff
@@ -0,0 +1,157 @@
+    }
+
+    JsonStreamBuffer(final Reader reader, final int initialBufferSize) {
+        this.initialBufferSize = initialBufferSize;
+        this.reader = reader;
+        resetBuffer();
+    }
+
+    public int getPosition() {
+        return position;
+    }
+
+    public int read() {
+        if (eof) {
+            throw new JsonParseException("Trying to read past EOF.");
+        }
+
+        // if we just unread, we need to use the last character read since it may not be in the
+        // buffer
+        if (reuseLastChar) {
+            reuseLastChar = false;
+            int reusedChar = lastChar;
+            lastChar = -1;
+            position++;
+            return reusedChar;
+        }
+
+        // use the buffer until we catch up to the stream position
+        if (position - bufferStartPos < bufferCount) {
+            int currChar = buffer[position - bufferStartPos];
+            lastChar = currChar;
+            position++;
+            return currChar;
+        }
+
+        if (markedPositions.isEmpty()) {
+            resetBuffer();
+        }
+
+        // otherwise, try and read from the stream
+        try {
+            int nextChar = reader.read();
+            if (nextChar != -1) {
+                lastChar = nextChar;
+                addToBuffer((char) nextChar);
+            }
+            position++;
+            if (nextChar == -1) {
+                eof = true;
+            }
+            return nextChar;
+
+        } catch (final IOException e) {
+            throw new JsonParseException(e);
+        }
+    }
+
+    private void resetBuffer() {
+        bufferStartPos = -1;
+        bufferCount = 0;
+        buffer = new char[initialBufferSize];
+    }
+
+    public void unread(final int c) {
+        eof = false;
+        if (c != -1 && lastChar == c) {
+            reuseLastChar = true;
+            position--;
+        }
+    }
+
+    public int mark() {
+        if (bufferCount == 0) {   // Why not markedPositions.isEmpty()?
+            bufferStartPos = position;
+        }
+        if (!markedPositions.contains(position)) {
+            markedPositions.add(position);
+        }
+        return position;
+    }
+
+    public void reset(final int markPos) {
+        if (markPos > position) {
+            throw new IllegalStateException("mark cannot reset ahead of position, only back");
+        }
+        int idx = markedPositions.indexOf(markPos);
+        if (idx == -1) {
+            throw new IllegalArgumentException("mark invalidated");
+        }
+        if (markPos != position) {
+            reuseLastChar = false;
+        }
+        markedPositions.subList(idx, markedPositions.size()).clear();
+        position = markPos;
+    }
+
+    public void discard(final int markPos) {
+        int idx = markedPositions.indexOf(markPos);
+        if (idx == -1) {
+            return;
+        }
+        // TODO: is this really safe? Can we assume that discarding an earlier mark also
```

Comment:
Any actions on these TODOs?

Reply:
I think it's fine now, since we haven't exposed a public discard method on BsonReaderMark.  I'll open a ticket for that and remove this TODO.

Reply:
Removed TODO and opened ticket